### PR TITLE
Integrate permits for getproof

### LINKED ISSUE #9298: Integrate Permits in EthAPI
### Describe the feature

with ethProof the EthApi now has an expensive endpoint that should be ratelimited.

the tracing endpoints, like debug already have a shared `BlockingTaskGuard`:

https://github.com/paradigmxyz/reth/blob/edbbc9636eca770102f6ff33779130deb7b38360/crates/rpc/rpc/src/debug.rs#L1024-L1030

However, these are aimed at the CPU, while eth_proof focuses more on the memory.
we can solve this with an additional guard in ethAPI itself. we want to explicitly restrict eth_getproof without interfering too much with the regular tracing calls.

## TODO

* integrate guard for ethProof in ethAPI

This needs
1) a new setting `proof_permits: usize` in the new function:
https://github.com/paradigmxyz/reth/blob/edbbc9636eca770102f6ff33779130deb7b38360/crates/rpc/rpc/src/eth/core.rs#L40-L47
2) add a guard instance: 
https://github.com/paradigmxyz/reth/blob/edbbc9636eca770102f6ff33779130deb7b38360/crates/rpc/rpc/src/eth/core.rs#L156-L156
3) expose acquire functions in https://github.com/paradigmxyz/reth/blob/edbbc9636eca770102f6ff33779130deb7b38360/crates/rpc/rpc-eth-api/src/helpers/blocking_task.rs#L10-L10
similar to
https://github.com/paradigmxyz/reth/blob/edbbc9636eca770102f6ff33779130deb7b38360/crates/tasks/src/pool.rs#L31-L39

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kostekIV**
Hi @mattsse, this is wip as I am not sure if this the right direction.

Closes #9298

### CODE REVIEW

File: crates/rpc/rpc-eth-api/src/helpers/blocking_task.rs [hunk 1]
```diff
@@ -18,6 +20,12 @@ pub trait SpawnBlocking: Clone + Send + Sync + 'static {
     /// Thread pool access in default trait method implementations.
     fn tracing_task_pool(&self) -> &BlockingTaskPool;
 
+    /// See also [Semaphore::acquire_owned](`tokio::sync::Semaphore::acquire_owned`).
+    async fn acquire_owned(&self) -> Result<OwnedSemaphorePermit, AcquireError>;
+
+    /// See also  [Semaphore::acquire_many_owned](`tokio::sync::Semaphore::acquire_many_owned`).
+    async fn acquire_many_owned(&self, n: u32) -> Result<OwnedSemaphorePermit, AcquireError>;
```
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/blocking_task.rs:27]: we can also return impl Future here and don't need async_trait, see `spawn_blocking_io` for ref

**Review by mattsse — CHANGES_REQUESTED** [review]
nice this is exactly what we want here.

integrating this in cli also makes sense


I think eventually we can combine the various kinds of permits into one type, tracing,proof but we can do that separately.

File: crates/rpc/rpc-eth-api/src/helpers/blocking_task.rs (hunk 1 above)
Comment (kostekIV) [reply crates/rpc/rpc-eth-api/src/helpers/blocking_task.rs]: Done

**Comment by kostekIV** [comment]
> nice this is exactly what we want here.
> 
> integrating this in cli also makes sense
> 
> I think eventually we can combine the various kinds of permits into one type, tracing,proof but we can do that separately.

Sure, I could do it here but maybe it would be slightly better to do in the next pr not to bloat this one?

**Review by mattsse — APPROVED** [review]
great! tysm

> better to do in the next pr not to bloat this one

definitely

**Comment by kostekIV** [comment]
> great! tysm
> 
> > better to do in the next pr not to bloat this one
> 
> definitely

Hey @mattsse, circling back on this, I'd said I'd do a follow-up PR and then totally forgot, sorry! took a quick look and I believe it still seems relevant, so happy to do it finally if you agree

**Pull request merged**